L22-cache/src/main/java/ru/otus/cachehw/MyCache.java

```diff
@@ -1,30 +1,70 @@
 package ru.otus.cachehw;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.WeakHashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 public class MyCache<K, V> implements HwCache<K, V> {
-    // Надо реализовать эти методы
+
+    private final WeakHashMap<K, V> cacheMap;
+
+    private final List<HwListener<K, V>> listenerList;
+
+    private static final Logger logger = LoggerFactory.getLogger(MyCache.class);
+
+    public MyCache() {
+        this.cacheMap = new WeakHashMap<>();
+        this.listenerList = new ArrayList<>();
+    }
 
     @Override
     public void put(K key, V value) {
-        throw new UnsupportedOperationException();
+        if (!cacheMap.containsKey(key)) {
+            cacheMap.put(key, value);
+            notifyListeners(key, value, CacheListenerMessage.KEY_VALUE_ADDED);
+        } else {
+            cacheMap.replace(key, value);
+            notifyListeners(key, value, CacheListenerMessage.KEY_VALUE_REPLACED);
+        }
+    }
+
+    private void notifyListeners(K key, V value, CacheListenerMessage cacheListenerMessage) {
+        for (HwListener<K, V> listener : listenerList) {
+            try {
+                listener.notify(key, value, cacheListenerMessage.getMessage());
+            } catch (Exception e) {
+                logger.error("Exception while notifying listener");
```

Comment:
```suggestion
                logger.error("Exception while notifying listener", e);
```

Не забывайте исключение отправлять в логи или, хотя бы, `e.getMessage()`.
Чтобы в логах было больше информации для диагностики ошибки.